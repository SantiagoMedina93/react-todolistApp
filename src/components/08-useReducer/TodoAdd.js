import React from 'react';
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {
    
    

    const [{ description }, handelInputChance, reset] = useForm({
        description:''
    }); 
    
    const handleSubmit = (event) => {
        event.preventDefault();

        if ( description.trim().length <= 1 ) {
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        }

        handleAddTodo( newTodo );
        reset();
    }


    
    return (
        <form onSubmit={ handleSubmit }>
            <input
                type='text'
                name='description'
                className='form-control'
                placeholder='Aprender.....'
                autoComplete='off'
                value={ description }
                onChange={ handelInputChance }
            />

            <button
                type='submit'
                className='btn btn-primary col-12 mt-2'
            >
                Agregar
            </button>

        </form>
    )
};