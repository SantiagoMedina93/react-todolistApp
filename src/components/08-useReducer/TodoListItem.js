import React from 'react';

export const TodoListItem = ({ todos, handleDelete, handleComplete,index}) => {
    
    return (
        <li
            key={ todos.id }
            className='list-group-item'
        >
            <p 
                className={ `${todos.done && 'complete'}` }
                onClick={ () => handleComplete( todos.id ) }
            >
                { index + 1 }. {todos.desc}
            </p>
            
            <button
                    onClick={ () => handleDelete( todos.id ) }
                    /* onClick={ handleDelete }
                    value={ id } */
                    className='btn btn-danger'
            >
                Borrar
            </button>
        </li>
    )
};