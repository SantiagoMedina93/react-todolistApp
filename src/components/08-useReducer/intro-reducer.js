const initialState = [{
    id:1,
    todo: 'Comprar pan',
    done: false
}];

const newTodo = [{
    id:2,
    todo: 'Comprar leche',
    done: false
}];

const agregarToDoAction = {
    type: 'agregar',
    payload: newTodo
}

const todoReducer = ( state = initialState, action ) => {

    if (action?.type === 'agregar') {
        return [...state, action.payload]
    }

    return state;
}


let todos = todoReducer();
todos = todoReducer( todos,agregarToDoAction );

console.log( todos );