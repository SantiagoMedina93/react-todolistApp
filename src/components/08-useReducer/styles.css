*{
    margin: 0px;
    padding: 0px;
    box-sizing: border-box;
}

body{ 
    padding: 50px;
}

p{
    margin: 0px;
    cursor: pointer;
}

.list-group-item{
    display: flex;
    align-items: center;
    justify-content: space-between;
}

.complete{
    text-decoration: line-through;
}