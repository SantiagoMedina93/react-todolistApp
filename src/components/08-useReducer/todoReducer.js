
export const todoReducer = ( state = [], {type, payload} ) => {
    
    switch (type) {
        case 'add':
            return [ ...state, payload ]

        case 'delete': 
            return state.filter( (state) => state.id !==  payload );
    
        case 'done-old': 
            return state.map( (state) => {
                if (state.id === payload) {
                    console.log(state);
                    return {
                        ...state,
                        done: !state.done
                    } 
                }else{
                    return state;
                }
            })

        case 'done': 

            return state.map( (state) => 
                ( state.id === payload )
                    ?{...state, done: !state.done}
                    : state
            )
            
        // Cuando se renderiza por primera vez, la fn todoReducer toma el caso default
        default:
            return state
        
    }

};