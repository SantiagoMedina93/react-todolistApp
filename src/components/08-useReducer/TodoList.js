import React from 'react';
import { TodoListItem } from './TodoListItem';

export const TodoList = ({todos, handleDelete, handleComplete}) => {

    
    return (
        <ul className='list-group list-group-flush'>     
            {
                todos.map( (todos,index) => (
                    
                    /* TodoListItem, todo, i, handleDelete, handelDone  */
                    <TodoListItem
                        key={ todos.id } 
                        todos={todos} 
                        index={index} 
                        handleComplete={ handleComplete } 
                        handleDelete={ handleDelete }
                    />
                ))
            }

        </ul>
    )
};