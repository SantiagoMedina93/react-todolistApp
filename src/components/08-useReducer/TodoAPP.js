import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';
import { useForm } from '../../hooks/useForm';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';

import './styles.css'

// Para pocos cambios de estados: useState
// Para muchos cambios de estado: useReducer


const init = () => {
    // localStorage.getItem(key) = null <==> key no existe
    // JSON.parse(null) = null
    return JSON.parse(localStorage.getItem('todos')) || [];


    /* return[{
        id: new Date().getTime(),
        desc: 'Aprender React',
        done: false
    }] */
}

/*  const initialState = [{
    id: new Date().getTime(),
    desc: 'Aprender Mongo',
    done: false
}] */


export const TodoAPP = () => {

    // State: Estado del hook
    // todoReducer = ( state = [], action ) => {}: Fn que switchea como cambia el estado
    // distach ( action ): Envia el argumento a todoReducer. Si cambia el state: Redibuja el componente

    const [ todos, dispatch ] = useReducer(todoReducer, [], init );

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos))
    }, [todos]);
 

    const handleDelete = (todoId) =>{

        const action = {
            type: 'delete',
            payload: todoId
        }

        dispatch( action );
    }

    const handleComplete = (id) => {

        dispatch( {
            type: 'done',
            payload: id
        } );
        
    }

    const handleAddTodo = (newTodo) => {

        dispatch({
            type: 'add',
            payload: newTodo
        });

    }

    return (
        <>
            <h1>TodoApp: <small>( {todos.length} )</small> </h1>
            <hr />

            <div className='row'>
                <div className='col-7'>
                    <TodoList todos={todos} handleComplete={ handleComplete } handleDelete={ handleDelete }/>
                </div>

                <div className='col-5'>
                    <h3 className='text-center'>Agregar ToDo</h3>
                    <hr />

                    <TodoAdd handleAddTodo={ handleAddTodo } /> 
                </div>
            </div>
            
        </>
    )
};