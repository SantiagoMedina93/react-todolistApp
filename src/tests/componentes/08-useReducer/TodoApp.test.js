import React from "react";
import { mount, shallow } from "enzyme";
import { TodoAPP } from '../../../components/08-useReducer/TodoAPP';
import { demoTodos } from '../../fixture/demoTodos'


import { act } from '@testing-library/react-hooks'

describe('Test in <TodoApp.test />', () => {

    const wrapper = shallow( <TodoAPP /> );
    Storage.prototype.setItem = jest.fn()

    test('Debe de mostrarse correctamnete', () => {

        expect( wrapper ).toMatchSnapshot();

    });

    test('Debe de agregar un TODO', () => {
    
        const wrapper = mount( <TodoAPP /> ); //Lo mismo que el shallow pero mas especifico
        
        act(() => {

            wrapper.find('TodoAdd').prop( 'handleAddTodo' )(demoTodos[0]);
            wrapper.find('TodoAdd').prop( 'handleAddTodo' )(demoTodos[1]);

        })

        expect( wrapper.find('h1').text().trim() ).toBe('TodoApp: ( 2 )');
        expect ( localStorage.setItem ).toBeCalledTimes(2);

    });

    test('Debe de eliminar un todo', () => {
    
        wrapper.find('TodoAdd').prop('handleAddTodo')( demoTodos[0] );
        wrapper.find('TodoList').prop('handleDelete')( demoTodos[0].id );

        expect( wrapper.find('h1').text().trim() ).toBe('TodoApp: ( 0 )');

        
    });
});