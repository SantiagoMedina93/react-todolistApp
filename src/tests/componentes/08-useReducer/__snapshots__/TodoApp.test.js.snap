// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <TodoApp.test /> Debe de mostrarse correctamnete 1`] = `
<Fragment>
  <h1>
    TodoApp: 
    <small>
      ( 
      0
       )
    </small>
     
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleComplete={[Function]}
        handleDelete={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <h3
        className="text-center"
      >
        Agregar ToDo
      </h3>
      <hr />
      <TodoAdd
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</Fragment>
`;
